### PR TITLE
Add `std.time.nanoTimestamp` function

### PR DESCRIPTION
**Pull request by leroycep**
Implementing this for #4635

### CODE REVIEW

**Comment by daurnimator** [comment]
You should be able to use `Once` (https://github.com/ziglang/zig/pull/5094)

**Comment by leroycep** [comment]
> You should be able to use `Once` (#5094)
Ah, okay. Thanks!

File: lib/std/time.zig [hunk 1]
```diff
@@ -51,25 +85,28 @@ pub fn milliTimestamp() u64 {
...
+        return ns;
     }
     if (comptime std.Target.current.isDarwin()) {
-        var tv: os.darwin.timeval = undefined;
-        var err = os.darwin.gettimeofday(&tv, null);
-        assert(err == 0);
-        const sec_ms = tv.tv_sec * ms_per_s;
-        const usec_ms = @divFloor(tv.tv_usec, us_per_s / ms_per_s);
-        return @intCast(u64, sec_ms + usec_ms);
+        // https://stackoverflow.com/a/21352348
```
Comment (daurnimator) [inline lib/std/time.zig:91]: I'm confused why we need this at all. 
OSX 10.13+ has a working `CLOCK_REALTIME`. 
This code is only required to support <= OSX 10.12 (which is unsupported by zig in general).

I guess we can keep it for older OSX, in which case we should put it behind a version check:
```zig
if (switch(std.Target.current.os.tag) {
    .macosx => std.Target.current.os.version_range.semver.min.major == 10 and std.Target.current.os.version_range.semver.min.minor < 12,
    // TODO: figure out which version CLOCK_REALTIME was added,
    .ios, .watchos, .tvos => false,
    else => false,
}) {
```

File: lib/std/time.zig [hunk 2]
```diff
@@ -33,16 +33,50 @@ pub fn timestamp() u64 {
...
+const DarwinTimeStart = struct {
+    timebase: os.darwin.mach_timebase_info_data,
+    inittime: os.darwin.timespec,
+    initclock: u64,
+};
+
+var _timestart: DarwinTimeStart = undefined;
+var _init_timestart_once = std.once(_init_timestart);
+
+pub fn _init_timestart() void {
```
Comment (daurnimator) [inline lib/std/time.zig:48]: probably doesnt need to be `pub`? also remove the `_`.

File: lib/std/time.zig [hunk 3]
```diff
@@ -33,16 +33,50 @@ pub fn timestamp() u64 {
...
+    return @divFloor(nanoTimestamp(), millisecond);
+}
+
+const DarwinTimeStart = struct {
+    timebase: os.darwin.mach_timebase_info_data,
+    inittime: os.darwin.timespec,
+    initclock: u64,
+};
+
+var _timestart: DarwinTimeStart = undefined;
```
Comment (daurnimator) [inline lib/std/time.zig:45]: no need for the `_` prefix.

File: lib/std/time.zig [hunk 4]
```diff
@@ -33,16 +33,50 @@ pub fn timestamp() u64 {
...
+const DarwinTimeStart = struct {
+    timebase: os.darwin.mach_timebase_info_data,
+    inittime: os.darwin.timespec,
+    initclock: u64,
+};
+
+var global_timestart: DarwinTimeStart = undefined;
+var init_global_timestart_once = std.once(init_global_timestart);
+
+pub fn init_global_timestart() void {
```
Comment (daurnimator) [inline lib/std/time.zig:48]: I think this shouldn't be `pub`? If we want to permit people to pre-load it then make `init_global_timestart_once` `pub`.

File: lib/std/time.zig [hunk 5]
```diff
@@ -33,16 +33,50 @@ pub fn timestamp() u64 {
...
+    timestart.initclock = os.darwin.mach_absolute_time();
+    timestart.inittime.tv_sec = micro.tv_sec;
+    timestart.inittime.tv_nsec = micro.tv_usec * 1000;
+
+    global_timestart = timestart;
+}
+
+/// Get the posix timestamp, UTC, in nanoseconds
+///
+/// On windows this only has a granularity of 100 nanoseconds.
```
Comment (daurnimator) [inline lib/std/time.zig:66]: FYI posix has `clock_getres` to indicate the granularity of the timer. We should probably wrap that too and mention it in this comment.

**Pull request closed without merging**